test/helpers/foundry/foundryup.ts

```diff
@@ -0,0 +1,166 @@
+import { Architecture, Checksums, Extension, Platform } from './types';
+
+export function getCacheDirectory(): string {
+  let enableGlobalCache = false;
+  try {
+    const configFileContent = readFileSync('.yarnrc.yml', 'utf8');
+    const parsedConfig = parseYaml(configFileContent);
+    enableGlobalCache = parsedConfig?.enableGlobalCache ?? false;
+  } catch (error) {
+    console.error('Error reading/parsing .yarnrc.yml:', error);
+  }
+  return enableGlobalCache
+    ? join(homedir(), '.cache', 'metamask')
+    : join(cwd(), '.metamask', 'cache');
+}
+
+export function getBinaryArchiveUrl(
+  repo: string,
+  tag: string,
+  version: string,
+  platform: Platform,
+  arch: string,
+): string {
+  const ext = platform === Platform.Windows ? Extension.Zip : Extension.Tar;
+  return `https://github.com/${repo}/releases/download/${tag}/foundry_${version}_${platform}_${arch}.${ext}`;
+}
+
+export async function checkAndDownloadBinaries(
+  url: URL,
+  binaries: string[],
+  cachePath: string,
+  platform: Platform,
+  arch: Architecture,
+  checksums?: Checksums,
+): Promise<Dir> {
+  let downloadedBinaries: Dir;
+  try {
+    say(`checking cache`);
+    downloadedBinaries = await opendir(cachePath);
+    say(`found binaries in cache`);
+  } catch (e: unknown) {
+    say(`binaries not in cache`);
+    if ((e as NodeJS.ErrnoException).code === 'ENOENT') {
+      say(`installing from ${url.toString()}`);
+      // directory doesn't exist, download and extract
+      const platformChecksums = transformChecksums(checksums, platform, arch);
+      await extractFrom(url, binaries, cachePath, platformChecksums);
+      downloadedBinaries = await opendir(cachePath);
+    } else {
+      throw e;
+    }
+  }
+  return downloadedBinaries;
+}
+
+export async function installBinaries(
+  downloadedBinaries: Dir,
+  BIN_DIR: string,
+  cachePath: string,
+): Promise<void> {
+  for await (const file of downloadedBinaries) {
+    if (!file.isFile()) {
+      continue;
+    }
+    const target = join(file.parentPath, file.name);
+    const path = join(BIN_DIR, relative(cachePath, target));
+    // clean up any existing files or symlinks
+    await unlink(path).catch(noop);
+    try {
+      // create new symlink
+      await symlink(target, path);
+    } catch (e) {
+      if (!(isCodedError(e) && ['EPERM', 'EXDEV'].includes(e.code))) {
+        throw e;
+      }
+      // symlinking can fail if it's a cross-device/filesystem link, or for
+      // permissions reasons, so we'll just copy the file instead
+      await copyFile(target, path);
+    }
+    // check that it works by logging the version
+    say(`installed - ${getVersion(path)}`);
+  }
+}
+
+export async function downloadAndInstallFoundryBinaries(): Promise<void> {
+  const parsedArgs = parseArgs();
+
+  const CACHE_DIR = getCacheDirectory();
+
+  if (parsedArgs.command === 'cache clean') {
+    await rm(CACHE_DIR, { recursive: true, force: true });
```

Comment:
Having a `cache clean` was mostly just useful when writing this code and manually testing it. Feature could be removed if we don't want to keep code we hopefully won't need.